File: java/ShellSort.java
public class ShellSort {
    public <T extends Comparable<T>> void sort(T[] arr) {
        int n = arr.length;

        for (int gap = n/2; gap > 0; gap /= 2) {
            for (int i = gap; i < n; i++) {
                T temp = arr[i];
                int j;
                for (j = i; j >= gap && arr[j - gap].compareTo(temp) > 0; j -= gap) {
                    arr[j] = arr[j - gap];
                }
                arr[j] = temp;
            }
        }
    }
}

File: src/shell_sort.rs
pub fn shell_sort<T: PartialOrd + Copy>(arr: &mut [T]) {
    let n = arr.len();

    // Use Marcin Ciura's gap sequence for optimal performance
    let gaps = [701, 301, 132, 57, 23, 10, 4, 1];

    for &gap in gaps.iter() {
        let mut i = gap;
        while i < n {
            let mut j = i;
            while j >= gap && arr[j - gap] > arr[j] {
                arr.swap(j, j - gap);
                j -= gap;
            }
            i += 1;
        }
    }
}